pytorch_lightning/profiler/profilers.py

```diff
@@ -543,3 +570,261 @@ def __del__(self):
         """Close profiler's stream."""
         if self.output_file:
             self.output_file.close()
+
+
+class ScheduleWrapper:
+    """
+    This class is used to override the schedule logic from the profiler and perform
+    recording for both `training_step`, `validation_step`.
+    """
+
+    def __init__(self, schedule: Callable):
+        self._schedule = schedule
+        self._num_training_step_and_backward = 0
+        self._num_validation_step = 0
+        self._training_step_and_backward_reached_end = False
+        self._validation_step_reached_end = False
+        # used to stop profiler when `ProfilerAction.RECORD_AND_SAVE` is reached.
+        self._none_action = None
+        self._current_action = None
+
+    @property
+    def num_step(self):
+        if self._current_action == "training_step_and_backward":
+            return self._num_training_step_and_backward
+        elif self._current_action == "validation_step":
+            return self._num_validation_step
+        elif self._current_action == "on_fit_start":
+            return 0
+        else:
+            pass
+
+    def _step(self):
+        if self._current_action == "training_step_and_backward":
+            self._num_training_step_and_backward += 1
+        elif self._current_action == "validation_step":
+            # skip sanity check
+            if self._num_training_step_and_backward > 0:
+                self._num_validation_step += 1
+        elif self._current_action == "on_fit_start":
+            pass
+
+    @property
+    def has_finished(self):
+        if self._current_action == "training_step_and_backward":
+            return self._training_step_and_backward_reached_end
+        elif self._current_action == "validation_step":
+            return self._validation_step_reached_end
+        else:
+            return False
+
+    def __call__(self, num_step: int):
+        # ignore the provided input. Keep internal state instead.
+        if self.has_finished:
+            return ProfilerAction.NONE
+
+        self._step()
+        action = self._schedule(self.num_step)
+        if action == ProfilerAction.RECORD_AND_SAVE:
+            if self._current_action == "training_step_and_backward":
+                self._training_step_and_backward_reached_end = True
+            elif self._current_action == "validation_step":
+                self._validation_step_reached_end = True
+        return action
+
+
+if _TORCH_GREATER_EQUAL_1_8:
```

Comment:
how will this influence docs?
If docs is build with v1.7, but user has v1.8 they will see the documentation for another version than what they are getting right? IMO we should be very clear somewhere that the `PytorchProfiler` is different depending on the users version.

Reply:
This was my point with https://github.com/PyTorchLightning/pytorch-lightning/pull/6067#issuecomment-786658875

Reply:
I pushed an updated version.